### ISSUE TITLE
Warn user about contact leakage before creating a large group

### ISSUE THREAD

**Issue by konklone** [WhisperSystems/TextSecure#3928]
TextSecure makes it too easy to accidentally leak your contacts' phone numbers to each other. It should make it more clear to the user -- perhaps through **a user warning if a group contains more than X people** -- that a group message to a large group will share everyone's phone numbers with everyone else.

Background: I've been an enthusiastic TextSecure user for a couple years now, but I hadn't done much group messaging. I reset my phone and so my TextSecure key had changed. I wanted to let everyone know en masse, and went to create a group of all my TextSecure contacts. Once I created the group (before I explicitly sent a message), it had already created the new group on all of my friends' phones, and I quickly learned that they could all see every other person's phone number.

Obviously, if I'd done much group messaging before on TextSecure, I would have known this. But I hadn't yet, and this was a rough way to learn it that resulted in me weakening the privacy of a good number of my friends.

I can imagine many solutions to address this, but the one that I think would involve the least engineering effort would be to warn a user that phone numbers will be shared before creating a group where the group is above a certain threshold. This could take the form of an OK/Cancel popup after hitting the button to create a group, or something more integrated.

Whatever the solution, I believe adding friction to mass-contact-leakage is well within TextSecure's commitment to a UX that makes secure communication easy. With a high enough threshold, it's unlikely to affect many users, and will only add that friction to a small number of important cases where that friction is constructive.

**Comment by waldoj** [comment]
+1

I was one of the recipients of this message, and did not like seeing my phone number exposed to a bunch of strangers. It was a confusing interaction, and I assumed that TextSecure—or at least @konklone's account—had been compromised in some way. Now I feel nervous about using TextSecure, wondering if I'll commit a similar gaffe.

**Comment by wslack** [comment]
+1 - to make things worse, if the user sees this and leaves the group immediately, there is no way to send a follow-up explanation.

**Comment by moxie0** [comment]
Thanks but I don't think we're going to do this.

**Comment by konklone** [comment]
@moxie0 You're under no moral obligation to agree with me or explain why you disagree, but you should know that responses like that are seriously demoralizing. I'm nnow a lot less likely to offer constructive feedback here.

I put a decent amount of effort into this ticket after I had a bad TextSecure experience (descending from my own mistake), so that TextSecure could become a better product and so fewer people would end up in my situation. You can disagree, but I was hoping my issue would be treated with more consideration than that.

**Comment by moxie0** [comment]
@konklone Sorry but this is an issue tracker, not a discussion forum.  All kinds of people open all kinds of issues, ranging from Q&A to rants to feature requests to bugs they encounter.  I read all of them, and I have to decide which of them realistically represent work which we are going to do.  If it's not work which we're going to do, I close the issue.

You've outlined your lack of familiarity with a feature, which is not TextSecure-specific, but how just about every other messenger works.  If you want to have a conversation about how to subtly change the way this feature is designed so that it is more clearly understood by people who have never used group messaging before, please feel free to throw those ideas out on the mailing list.  But you've opened an issue for a popup, and that's not something we're going to do -- we need to remove the dialogs we have now, not create new ones.

Since this aspires to be an issue tracker for work we're going to do, and this outlines work which doesn't meet our objectives, I've closed the issue.

**Comment by wslack** [comment]
> how just about every other messenger works

@moxie0 I've only used a few other messenger apps, but in my experience this was different in a few ways from my past experience:

1. The group was formed in the app before any first message was sent; ie, without context.
2. Every user listed in the blast received a separate individual conversation in the middle of my inbox (either when I left the group convo or before)

I am happy to open separate issues for each of those if that would be helpful.

Sources: http://joyofandroid.com/how-to-send-group-texts-from-android-devices/
https://gigaom.com/2011/11/01/imessage-101-group-messaging/

**Comment by konklone** [comment]
@moxie0 I understand this isn't a discussion forum, this is a comment you've made on numerous occasions. This ticket meets my understanding of a specific feature request, which is appropriate on most issue trackers I'm familiar with.

I opened an issue for something which could be a popup, but doesn't have to be. Warnings can also be inline and provide no popups or user friction. If there's something to be discussed here

> You've outlined your lack of familiarity with a feature, which is not TextSecure-specific, but how just about every other messenger works.

I'm very familiar with group messaging. I've used it in traditional SMS and messaging apps before. I still did not expect TextSecure to exhibit this behavior. This could certainly just be me being thoughtless, but it still may be something worth TextSecure's consideration.

As @wslack notes, TextSecure has a piece of unusual behavior, which is that it blasts out to users before an actual message is sent. I can understand why the protocol might demand something like this, but it does mean that TextSecure makes it easier for someone to get into this situation -- I don't have any chance to think twice between creating a group and using the group. The two actions are the same.

> If you want to have a conversation about how to subtly change the way this feature is designed so that it is more clearly understood by people who have never used group messaging before, please feel free to throw those ideas out on the mailing list.

OK, I'll do that.
